src/Data/Vector/Hashtables/Internal.hs

```diff
@@ -286,29 +312,297 @@ instance DeleteEntry B.MVector where
+  | i >= 0 = do
+      hc <- hashCode ! i
+      k <- key ! i
+      if hc == hashCode' && k == key' then do
+          nxt <- next ! i
+          if last < 0
+              then buckets <~ bucket $ nxt
+              else next <~ last $ nxt
+          hashCode <~ i $ -1
+          next <~ i =<< refs ! getFreeList
+          deleteEntry key i
+          deleteEntry value i
+          refs <~ getFreeList $ i
+          fc <- refs ! getFreeCount
+          refs <~ getFreeCount $ fc + 1
+        else deleteWithIndex bucket hashCode' d key' i =<< next ! i
+  | otherwise = return ()
+{-# INLINE deleteWithIndex #-}
+
+lookup :: (MVector ks k, MVector vs v, PrimMonad m, Hashable k, Eq k)
+   => Dictionary (PrimState m) ks k vs v -> k -> m (Maybe v)
+lookup = at'
+{-# INLINE lookup #-}
+
+lookup' :: (MVector ks k, MVector vs v, PrimMonad m, Hashable k, Eq k)
+   => Dictionary (PrimState m) ks k vs v -> k -> m v
+lookup' = at
+{-# INLINE lookup' #-}
+
+lookupIndex :: (MVector ks k, MVector vs v, PrimMonad m, Hashable k, Eq k)
+   => Dictionary (PrimState m) ks k vs v -> k -> m (Maybe Int)
+lookupIndex ht k = do
+    let safeIndex i | i < 0 = Nothing
+                    | otherwise = Just i
+    return . safeIndex =<< findEntry ht k
+{-# INLINE lookupIndex #-}
+
+null
+  :: (MVector ks k, PrimMonad m)
+  => Dictionary (PrimState m) ks k vs v -> m Bool
+null ht = return . (== 0) =<< length ht
+{-# INLINE null #-}
+
+length
+  :: (MVector ks k, PrimMonad m)
+  => Dictionary (PrimState m) ks k vs v -> m Int
+length DRef {..} = do
+    Dictionary {..} <- readMutVar getDRef
+    count           <- refs ! getCount
+    freeCount       <- refs ! getFreeCount
+    return (count - freeCount)
+{-# INLINE length #-}
+
+size
+  :: (MVector ks k, PrimMonad m)
+  => Dictionary (PrimState m) ks k vs v -> m Int
+size = length
+{-# INLINE size #-}
+
+member
+  :: (MVector ks k, MVector vs v, PrimMonad m, Hashable k, Eq k)
+  => Dictionary (PrimState m) ks k vs v -> k -> m Bool
+member ht k = return . (>= 0) =<< findEntry ht k
+{-# INLINE member #-}
+
+findWithDefault
+  :: (MVector ks k, MVector vs v, PrimMonad m, Hashable k, Eq k)
+  => Dictionary (PrimState m) ks k vs v -> v -> k -> m v
+findWithDefault ht v k = return . fromMaybe v =<< at' ht k
+{-# INLINE findWithDefault #-}
+
+alter
+  :: ( MVector ks k, MVector vs v, DeleteEntry ks, DeleteEntry vs
+     , PrimMonad m, Hashable k, Eq k
+     )
+  => Dictionary (PrimState m) ks k vs v -> (Maybe v -> Maybe v) -> k -> m ()
+alter ht f k = do
+  d@Dictionary{..} <- readMutVar . getDRef $ ht
+  let
+      hashCode' = hash k .&. 0x7FFFFFFFFFFFFFFF
+      targetBucket = hashCode' `rem` V.length buckets
+
+      onFound' value' dict i = insertWithIndex targetBucket hashCode' k value' (getDRef ht) dict i
+      onNothing' dict i = deleteWithIndex targetBucket hashCode' d k (-1) i
+
+      onFound dict i = do
+        d'@Dictionary{..} <- readMutVar . getDRef $ dict
+        v <- value ! i
+        case f (Just v) of
```

Comment:
I'd apply `maybe` here and for `onNothing` below (changing the order of argument to `onFound'` so that value goes last), but it's purely stylistic, I guess, so not important.